### PR TITLE
Sonic Config Mgmt Class

### PR DESCRIPTION
**Pull request by praveen-li**
**- What I did**
Provided a class in sonic_util config for 
-- Config Validation,
-- Delete Ports
-- Add Ports 

**- How I did it**
--Config_mgmt Class uploads config data from config_Db.json or from config db using SONiC-py-swsssdk APIs . 
--Port, vlan and acl related config is cropped out of entire config, this is because yang models are available only for Vlan, Acl and Port as of now. 
--Cropped portion of config DB will be given as input to Python Yang LIB while loading config data.
 
--Config_mgmt Class provides APIs to delete and add Ports. 
--Using Python yang LIB APIs, this class finds dependencies for each of the port which must be deleted and will delete dependencies. 
-- At last after successful validation, new config is written in config DB.
-- While addition, relevant default config is found for ports which are being added. Then entire new config [default + port] is added to current config in data tree.
-- At last after successful validation, new config is written in config DB.

**- How to verify it**
On Sonic Switch
```
Install 
Libyang, cpp, py2libyang, sonic-yang-mgmt
Pre requisite on sonic switch, need: default config


admin@lnos-x1-a-asw01:~/config_mgmt$ echo "On Sonic Switch"
On Sonic Switch
admin@lnos-x1-a-asw01:~/config_mgmt$ date
Wed Dec 11 01:53:37 UTC 2019
admin@lnos-x1-a-asw01:~/config_mgmt$ python config_mgmt.py
Test Run Delete Ports
Loaded below Yang Models
['sonic-port', 'sonic-acl', 'sonic-portchannel', 'sonic-vlan', 'sonic-interface', 'sonic-head']
Reading data from Redis configDb

Start Port Deletion
Find dependecies for port Ethernet0
Find dependecies for port Ethernet1
Find dependecies for port Ethernet2
Find dependecies for port Ethernet3
Deleting Port: Ethernet0
Deleting Port: Ethernet1
Deleting Port: Ethernet2
Deleting Port: Ethernet3
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

***Port Deletion Test Passed***


Start Port Addition
Generating default config for ['Ethernet0', 'Ethernet2']
Merge Port Config for ['Ethernet0', 'Ethernet2']
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

***Port Addition Test Passed***

admin@lnos-x1-a-asw01:~/config_mgmt$
```

[Pasted entire config, let me know If I need to attach config, instead of paste]
Logs:

```
pchaudha@7cc26115656c:/sonic/src/sonic-utilities/config$ python config_mgmt.py
Test Run Delete Ports
Loaded below Yang Models
['sonic-vlan', 'sonic-head', 'sonic-portchannel', 'sonic-acl', 'sonic-port', 'sonic-interface']
Reading data from config_db.json

Start Port Deletion
Deleting Port: Ethernet0
Deleting Port: Ethernet1
Deleting Port: Ethernet2
Deleting Port: Ethernet3
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

***Port Deletion Test Passed***


Start Port Addition
Generating default config for ['Ethernet0', 'Ethernet2']
Merge Port Config for ['Ethernet0', 'Ethernet2']
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

***Port Addition Test Passed***
```

Start Config:
```
{
    "VLAN_INTERFACE": {
        "Vlan100|2a04:5555:45:6709::/64": {
            "scope": "global", 
            "family": "IPv6"
        }, 
        "Vlan100|10.222.10.64/26": {
            "scope": "global", 
            "family": "IPv4"
        }, 
        "Vlan777|fe80::/10": {
            "scope": "local", 
            "family": "IPv6"
        }, 
        "Vlan777|10.111.58.64/26": {
            "scope": "global", 
            "family": "IPv4"
        }, 
        "Vlan777|2a04:5555:41:4e9::/64": {
            "scope": "global", 
            "family": "IPv6"
        }, 
        "Vlan100|fe80::/10": {
            "scope": "local", 
            "family": "IPv6"
        }
    }, 
    "ACL_RULE": {
        "V4-ACL-TABLE|Rule_20": {
            "SRC_IP": "10.222.0.0/15", 
            "PACKET_ACTION": "FORWARD", 
            "DST_IP": "10.222.72.0/26", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "777780"
        }, 
        "V4-ACL-TABLE|Rule_111": {
            "SRC_IP": "10.222.0.0/15", 
            "PACKET_ACTION": "FORWARD", 
            "DST_IP": "10.152.17.52/32", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "777700"
        }, 
        "V6-ACL-TBLE|Rule_80": {
            "PACKET_ACTION": "FORWARD", 
            "PRIORITY": "777720", 
            "SRC_IPV6": "2a04:5555:41::/48", 
            "DST_IPV6": "2a04:5555:43:341::/64", 
            "IP_TYPE": "IPv6ANY"
        }, 
        "V6-ACL-TBLE|Rule_60": {
            "PACKET_ACTION": "FORWARD", 
            "PRIORITY": "777740", 
            "SRC_IPV6": "2a04:5555:41::/48", 
            "DST_IPV6": "2a04:5555:43:340::/64", 
            "IP_TYPE": "IPv6ANY"
        }, 
        "V4-ACL-TABLE|DEFAULT_DENY": {
            "PACKET_ACTION": "DROP", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "0"
        }, 
        "V4-ACL-TABLE|Rule_180": {
            "SRC_IP": "10.222.0.0/15", 
            "PACKET_ACTION": "FORWARD", 
            "DST_IP": "10.252.222.21/32", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "777820"
        }, 
        "V4-ACL-TABLE|Rule_40": {
            "SRC_IP": "10.222.0.0/15", 
            "PACKET_ACTION": "FORWARD", 
            "DST_IP": "10.222.72.64/26", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "777760"
        }, 
        "V4-ACL-TABLE|Rule_120": {
            "SRC_IP": "10.222.0.0/15", 
            "PACKET_ACTION": "FORWARD", 
            "DST_IP": "10.252.208.41/32", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "777880"
        }, 
        "V6-ACL-TBLE|Rule_111": {
            "PACKET_ACTION": "FORWARD", 
            "PRIORITY": "777700", 
            "SRC_IPV6": "2a04:5555:41::/48", 
            "DST_IPV6": "2a04:5555:32:12::/64", 
            "IP_TYPE": "IPv6ANY"
        }, 
        "V4-ACL-TABLE|Rule_11100": {
            "SRC_IP": "0.0.0.0/0", 
            "PACKET_ACTION": "FORWARD", 
            "DST_IP": "0.0.0.0/0", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "990000"
        }, 
        "V4-ACL-TABLE|Rule_9000": {
            "SRC_IP": "10.222.0.0/15", 
            "PACKET_ACTION": "DROP", 
            "DST_IP": "0.0.0.0/0", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "991110"
        }, 
        "V4-ACL-TABLE|Rule_80": {
            "SRC_IP": "10.222.0.0/15", 
            "PACKET_ACTION": "FORWARD", 
            "DST_IP": "10.222.80.64/26", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "777720"
        }, 
        "V6-ACL-TBLE|Rule_9000": {
            "PACKET_ACTION": "DROP", 
            "PRIORITY": "991110", 
            "SRC_IPV6": "2a04:5555:41::/48", 
            "DST_IPV6": "::/0", 
            "IP_TYPE": "IPv6ANY"
        }, 
        "V6-ACL-TBLE|DEFAULT_DENY": {
            "PACKET_ACTION": "DROP", 
            "IP_TYPE": "IPv6ANY", 
            "PRIORITY": "0"
        }, 
        "V6-ACL-TBLE|Rule_40": {
            "PACKET_ACTION": "FORWARD", 
            "PRIORITY": "777760", 
            "SRC_IPV6": "2a04:5555:41::/48", 
            "DST_IPV6": "2a04:5555:43:321::/64", 
            "IP_TYPE": "IPv6ANY"
        }, 
        "V4-ACL-TABLE|Rule_60": {
            "SRC_IP": "10.222.0.0/15", 
            "PACKET_ACTION": "FORWARD", 
            "DST_IP": "10.222.80.0/26", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "777740"
        }, 
        "V4-ACL-TABLE|Rule_160": {
            "SRC_IP": "10.222.0.0/15", 
            "PACKET_ACTION": "FORWARD", 
            "DST_IP": "10.222.1.245/32", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "777840"
        }, 
        "V6-ACL-TBLE|Rule_20": {
            "PACKET_ACTION": "FORWARD", 
            "PRIORITY": "777780", 
            "SRC_IPV6": "2a04:5555:41::/48", 
            "DST_IPV6": "2a04:5555:43:320::/64", 
            "IP_TYPE": "IPv6ANY"
        }, 
        "V4-ACL-TABLE|Rule_140": {
            "SRC_IP": "10.222.0.0/15", 
            "PACKET_ACTION": "FORWARD", 
            "DST_IP": "10.148.128.245/32", 
            "IP_TYPE": "IPv4ANY", 
            "PRIORITY": "777860"
        }, 
        "V6-ACL-TBLE|Rule_11100": {
            "PACKET_ACTION": "FORWARD", 
            "PRIORITY": "990000", 
            "SRC_IPV6": "::/0", 
            "DST_IPV6": "::/0", 
            "IP_TYPE": "IPv6ANY"
        }
    }, 
    "VLAN": {
        "Vlan100": {
            "description": "svlan", 
            "dhcp_servers": [
                "10.222.72.116"
            ], 
            "vlanid": "111", 
            "mtu": "9216", 
            "admin_status": "up", 
            "members": [
                "Ethernet8", 
                "Ethernet6", 
                "Ethernet4", 
                "Ethernet5", 
                "Ethernet9", 
                "Ethernet7", 
                "Ethernet32", 
                "Ethernet30", 
                "Ethernet31", 
                "Ethernet36", 
                "Ethernet34", 
                "Ethernet33", 
                "Ethernet35", 
                "Ethernet29", 
                "Ethernet21", 
                "Ethernet20", 
                "Ethernet23", 
                "Ethernet22", 
                "Ethernet27", 
                "Ethernet26", 
                "Ethernet18", 
                "Ethernet19", 
                "Ethernet14", 
                "Ethernet15", 
                "Ethernet16", 
                "Ethernet17", 
                "Ethernet10", 
                "Ethernet11", 
                "Ethernet12", 
                "Ethernet13", 
                "Ethernet28", 
                "Ethernet0", 
                "Ethernet2"
            ]
        }, 
        "Vlan777": {
            "description": "pvlan", 
            "dhcp_servers": [
                "10.222.72.116"
            ], 
            "vlanid": "777", 
            "mtu": "9216", 
            "admin_status": "up", 
            "members": [
                "Ethernet9", 
                "Ethernet8", 
                "Ethernet27", 
                "Ethernet14", 
                "Ethernet35", 
                "Ethernet0", 
                "Ethernet2"
            ]
        }
    }, 
    "ACL_TABLE": {
        "V6-ACL-TBLE": {
            "type": "L3V6", 
            "policy_desc": "V6-ACL-TBLE", 
            "ports": [
                "Ethernet14", 
                "Ethernet15", 
                "Ethernet23", 
                "Ethernet30", 
                "Ethernet31", 
                "Ethernet18", 
                "Ethernet19", 
                "Ethernet25", 
                "Ethernet24"
            ]
        }, 
        "NO-NSW-PACL-V4": {
            "ports": [
                "Ethernet0", 
                "Ethernet2"
            ]
        }, 
        "NO-NSW-PACL-V6": {
            "ports": [
                "Ethernet0", 
                "Ethernet2"
            ]
        }, 
        "V4-ACL-TABLE": {
            "type": "L3", 
            "policy_desc": "V4-ACL-TABLE", 
            "ports": [
                "Ethernet26", 
                "Ethernet27", 
                "Ethernet24"
            ]
        }
    }, 
    "INTERFACE": {
        "Ethernet112|2a04:5555:40:a709::/126": {
            "scope": "global", 
            "family": "IPv6"
        }, 
        "Ethernet14|10.184.229.210/31": {
            "scope": "global", 
            "family": "IPv4"
        }, 
        "Ethernet16|10.184.230.210/31": {
            "scope": "global", 
            "family": "IPv4"
        }, 
        "Ethernet14|2a04:5555:40:a749::/126": {
            "scope": "global", 
            "family": "IPv6"
        }, 
        "Ethernet18|10.184.231.210/31": {
            "scope": "global", 
            "family": "IPv4"
        }, 
        "Ethernet16|2a04:5555:40:a789::/126": {
            "scope": "global", 
            "family": "IPv6"
        }, 
        "Ethernet112|10.184.228.210/31": {
            "scope": "global", 
            "family": "IPv4"
        }, 
        "Ethernet18|2a04:5555:40:a7c9::/126": {
            "scope": "global", 
            "family": "IPv6"
        }
    }, 
    "VLAN_MEMBER": {
        "Vlan100|Ethernet29": {
            "tagging_mode": "untagged"
        }, 
        "Vlan100|Ethernet2": {
            "tagging_mode": "untagged"
        }, 
        "Vlan100|Ethernet32": {
            "tagging_mode": "untagged"
        }, 
        "Vlan100|Ethernet33": {
            "tagging_mode": "untagged"
        }, 
        "Vlan100|Ethernet30": {
            "tagging_mode": "untagged"
        }, 
        "Vlan100|Ethernet31": {
            "tagging_mode": "untagged"
        }, 
        "Vlan100|Ethernet6": {
            "tagging_mode": "untagged"
        }, 
        "Vlan100|Ethernet4": {
            "tagging_mode": "untagged"
        }, 
        "Vlan100|Ethernet5": {
            "tagging_mode": "untagged"
        }, 
        "Vlan100|Ethernet36": {
            "tagging_mode": "untagged"
        }, 
        "Vlan777|Ethernet0": {
            "tagging_mode": "tagged"
        }, 
        "Vlan100|Ethernet34": {
            "tagging_mode": "untagged"
        }, 
        "Vlan777|Ethernet2": {
            "tagging_mode": "tagged"
        }, 
        "Vlan100|Ethernet35": {
            "tagging_mode": "untagged"
        }, 
        "Vlan100|Ethernet0": {
            "tagging_mode": "untagged"
        }
    }, 
    "PORT": {
        "Ethernet8": {
            "alias": "Eth3/1", 
            "admin_status": "up", 
            "lanes": "73", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet9": {
            "alias": "Eth3/2", 
            "admin_status": "up", 
            "lanes": "74", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet2": {
            "alias": "Eth1/3", 
            "admin_status": "up", 
            "lanes": "67, 68", 
            "description": "", 
            "speed": "50000"
        }, 
        "Ethernet0": {
            "alias": "Eth1/1", 
            "admin_status": "up", 
            "lanes": "65, 66", 
            "description": "", 
            "speed": "50000"
        }, 
        "Ethernet6": {
            "alias": "Eth2/3", 
            "admin_status": "up", 
            "lanes": "71", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet7": {
            "alias": "Eth2/4", 
            "admin_status": "up", 
            "lanes": "72", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet4": {
            "alias": "Eth2/1", 
            "admin_status": "up", 
            "lanes": "69", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet5": {
            "alias": "Eth2/2", 
            "admin_status": "up", 
            "lanes": "70", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet22": {
            "alias": "Eth6/3", 
            "admin_status": "up", 
            "lanes": "39", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet18": {
            "alias": "Eth5/3", 
            "admin_status": "up", 
            "lanes": "35", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet19": {
            "alias": "Eth5/4", 
            "admin_status": "up", 
            "lanes": "36", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet32": {
            "alias": "Eth9/1", 
            "admin_status": "up", 
            "lanes": "49", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet33": {
            "alias": "Eth9/2", 
            "admin_status": "up", 
            "lanes": "50", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet30": {
            "alias": "Eth8/3", 
            "admin_status": "up", 
            "lanes": "47", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet31": {
            "alias": "Eth8/4", 
            "admin_status": "up", 
            "lanes": "48", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet10": {
            "alias": "Eth3/3", 
            "admin_status": "up", 
            "lanes": "75", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet11": {
            "alias": "Eth3/4", 
            "admin_status": "up", 
            "lanes": "76", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet12": {
            "alias": "Eth4/1", 
            "admin_status": "up", 
            "lanes": "77", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet13": {
            "alias": "Eth4/2", 
            "admin_status": "up", 
            "lanes": "78", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet14": {
            "alias": "Eth4/3", 
            "admin_status": "up", 
            "lanes": "79", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet15": {
            "alias": "Eth4/4", 
            "admin_status": "up", 
            "lanes": "80", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet112": {
            "alias": "Eth29/1", 
            "admin_status": "up", 
            "lanes": "113,114", 
            "fec": "fc", 
            "description": "50G|dccsw01.nw|Eth18"
        }, 
        "Ethernet16": {
            "alias": "Eth5/1", 
            "admin_status": "up", 
            "lanes": "33", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet17": {
            "alias": "Eth5/2", 
            "admin_status": "up", 
            "lanes": "34", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet36": {
            "alias": "Eth10/1", 
            "admin_status": "up", 
            "lanes": "53", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet29": {
            "alias": "Eth8/2", 
            "admin_status": "up", 
            "lanes": "46", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet28": {
            "alias": "Eth8/1", 
            "admin_status": "up", 
            "lanes": "45", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet34": {
            "alias": "Eth9/3", 
            "admin_status": "up", 
            "lanes": "51", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet21": {
            "alias": "Eth6/2", 
            "admin_status": "up", 
            "lanes": "38", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet20": {
            "alias": "Eth6/1", 
            "admin_status": "up", 
            "lanes": "37", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet23": {
            "alias": "Eth6/4", 
            "admin_status": "up", 
            "lanes": "40", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet35": {
            "alias": "Eth9/4", 
            "admin_status": "up", 
            "lanes": "52", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet25": {
            "alias": "Eth7/2", 
            "admin_status": "up", 
            "lanes": "42", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet24": {
            "alias": "Eth7/1", 
            "admin_status": "up", 
            "lanes": "41", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet27": {
            "alias": "Eth7/4", 
            "admin_status": "up", 
            "lanes": "44", 
            "description": "", 
            "speed": "11100"
        }, 
        "Ethernet26": {
            "alias": "Eth7/3", 
            "admin_status": "up", 
            "lanes": "43", 
            "description": "", 
            "speed": "11100"
        }
    }
}
```

Config Diff after port deletion:
```
{
    "VLAN": {
        "Vlan100": {
            "members": [
                "Ethernet8", 
                "Ethernet6", 
                "Ethernet4", 
                "Ethernet5", 
                "Ethernet9", 
                "Ethernet7", 
                "Ethernet32", 
                "Ethernet30", 
                "Ethernet31", 
                "Ethernet36", 
                "Ethernet34", 
                "Ethernet33", 
                "Ethernet35", 
                "Ethernet29", 
                "Ethernet21", 
                "Ethernet20", 
                "Ethernet23", 
                "Ethernet22", 
                "Ethernet27", 
                "Ethernet26", 
                "Ethernet18", 
                "Ethernet19", 
                "Ethernet14", 
                "Ethernet15", 
                "Ethernet16", 
                "Ethernet17", 
                "Ethernet10", 
                "Ethernet11", 
                "Ethernet12", 
                "Ethernet13", 
                "Ethernet28"
            ]
        }, 
        "Vlan777": {
            "members": [
                "Ethernet9", 
                "Ethernet8", 
                "Ethernet27", 
                "Ethernet14", 
                "Ethernet35"
            ]
        }
    }, 
    "VLAN_MEMBER": {
        "Vlan100|Ethernet2": null, 
        "Vlan100|Ethernet3": null, 
        "Vlan100|Ethernet0": null, 
        "Vlan100|Ethernet1": null
    }, 
    "PORT": {
        "Ethernet2": null, 
        "Ethernet3": null, 
        "Ethernet0": null, 
        "Ethernet1": null
    }
}
```

Config Diff after port addition:
```
{
    "ACL_TABLE": {
        "NO-NSW-PACL-V4": {
            "ports": [
                "Ethernet0", 
                "Ethernet2"
            ]
        }, 
        "NO-NSW-PACL-V6": {
            "ports": [
                "Ethernet0", 
                "Ethernet2"
            ]
        }
    }, 
    "VLAN": {
        "Vlan100": {
            "members": [
                "Ethernet8", 
                "Ethernet6", 
                "Ethernet4", 
                "Ethernet5", 
                "Ethernet9", 
                "Ethernet7", 
                "Ethernet32", 
                "Ethernet30", 
                "Ethernet31", 
                "Ethernet36", 
                "Ethernet34", 
                "Ethernet33", 
                "Ethernet35", 
                "Ethernet29", 
                "Ethernet21", 
                "Ethernet20", 
                "Ethernet23", 
                "Ethernet22", 
                "Ethernet27", 
                "Ethernet26", 
                "Ethernet18", 
                "Ethernet19", 
                "Ethernet14", 
                "Ethernet15", 
                "Ethernet16", 
                "Ethernet17", 
                "Ethernet10", 
                "Ethernet11", 
                "Ethernet12", 
                "Ethernet13", 
                "Ethernet28", 
                "Ethernet0", 
                "Ethernet2"
            ]
        }, 
        "Vlan777": {
            "members": [
                "Ethernet9", 
                "Ethernet8", 
                "Ethernet27", 
                "Ethernet14", 
                "Ethernet35", 
                "Ethernet0", 
                "Ethernet2"
            ]
        }
    }, 
    "VLAN_MEMBER": {
        "Vlan100|Ethernet2": {
            "tagging_mode": "untagged"
        }, 
        "Vlan777|Ethernet2": {
            "tagging_mode": "tagged"
        }, 
        "Vlan777|Ethernet0": {
            "tagging_mode": "tagged"
        }, 
        "Vlan100|Ethernet0": {
            "tagging_mode": "untagged"
        }
    }, 
    "PORT": {
        "Ethernet2": {
            "alias": "Eth1/3", 
            "admin_status": "up", 
            "lanes": "67, 68", 
            "description": "", 
            "speed": "50000"
        }, 
        "Ethernet0": {
            "alias": "Eth1/1", 
            "admin_status": "up", 
            "lanes": "65, 66", 
            "description": "", 
            "speed": "50000"
        }
    }
}
```

Config Reload and load Test:
```

admin@lnos-x1-a-asw01:~/config_mgmt$ sudo config reload
Clear current config and reload config from the file /etc/sonic/config_db.json? [y/N]: y
Loaded below Yang Models
['sonic-port', 'sonic-acl', 'sonic-portchannel', 'sonic-vlan', 'sonic-interface', 'sonic-head']
Reading data from /etc/sonic/config_db.json
Data Validation successful
Running command: systemctl stop dhcp_relay

Running command: /usr/local/bin/sonic-cfggen -j orig_config_db.json --write-to-db
admin@lnos-x1-a-asw01:~/config_mgmt$ sudo config load orig_config_db.json -y -c
Loaded below Yang Models
['sonic-port', 'sonic-acl', 'sonic-portchannel', 'sonic-vlan', 'sonic-interface', 'sonic-head']
Reading data from orig_config_db.json
Data Validation successful
Running command: /usr/local/bin/sonic-cfggen -j orig_config_db.json --write-to-db
```

**- Previous command output (if the output of a command-line utility has changed)**
```
admin@lnos-x1-a-asw01:~/config_mgmt$ sudo config load orig_config_db.json -y
Running command: /usr/local/bin/sonic-cfggen -j orig_config_db.json --write-to-db
```
**- New command output (if the output of a command-line utility has changed)**
```
admin@lnos-x1-a-asw01:~/config_mgmt$ sudo config load orig_config_db.json -y -c
Loaded below Yang Models
['sonic-port', 'sonic-acl', 'sonic-portchannel', 'sonic-vlan', 'sonic-interface', 'sonic-head']
Reading data from orig_config_db.json
Data Validation successful
Running command: /usr/local/bin/sonic-cfggen -j orig_config_db.json --write-to-db
```

Testing after commit 7:
```
admin@lnos-x1-a-asw01:~/config_mgmt$ date
Fri Dec 13 00:48:00 UTC 2019
admin@lnos-x1-a-asw01:~/config_mgmt$ sudo config load orig_config_db.json -y -c
Loaded below Yang Models
['sonic-port', 'sonic-acl', 'sonic-portchannel', 'sonic-vlan', 'sonic-interface', 'sonic-head']
Reading data from orig_config_db.json
Data Validation successful
Running command: /usr/local/bin/sonic-cfggen -j orig_config_db.json --write-to-db

admin@lnos-x1-a-asw01:~/config_mgmt$ python config_mgmt.py
Test Run Delete Ports
Loaded below Yang Models
['sonic-port', 'sonic-acl', 'sonic-portchannel', 'sonic-vlan', 'sonic-interface', 'sonic-head']
Reading data from Redis configDb

Start Port Deletion
Find dependecies for port Ethernet0
Find dependecies for port Ethernet1
Find dependecies for port Ethernet2
Find dependecies for port Ethernet3
Deleting Port: Ethernet0
Deleting Port: Ethernet1
Deleting Port: Ethernet2
Deleting Port: Ethernet3
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

***Port Deletion Test Passed***


Start Port Addition
Generating default config for ['Ethernet0', 'Ethernet2']
Merge Port Config for ['Ethernet0', 'Ethernet2']
Data Validation successful
Generate Final Config to write in DB
Writing in Config DB

***Port Addition Test Passed***
```

### CODE REVIEW

File: config/config_mgmt.py [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+            # No further action with no force and deps exist
+            if force == False and deps:
+                return deps, False;
+
+            # delets all deps, No topological sort is needed as of now, if deletion
+            # of deps fails, return immediately
+            elif deps and force:
+                for dep in deps:
+                    # print(dep)
+                    self.sy.delete_node(str(dep))
```
Comment (praveen-li) [inline config/config_mgmt.py:149]: TODO: Return Exception from delete_node in _sonic_yang_ext.py

File: config/config_mgmt.py [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+        return True
+
+    """
+    Merge second dict in first, Note both first and second dict will be changed
+    First Dict will have merged part D1 + D2
+    Second dict will have D2 - D1 [unique keys in D2]
+    """
+    def mergeConfigs(self, D1, D2):
+
+        try:
```
Comment (praveen-li) [inline config/config_mgmt.py]: TODO: This try block can move down.

File: config/config_mgmt.py [hunk 3]
```diff
@@ -0,0 +1,583 @@
...
+    deps, ret = cm.deletePorts(delPorts=['Ethernet0', 'Ethernet1', 'Ethernet2', 'Ethernet3'], force=True)
+    if ret == False:
+        print("Port Deletion Test failed")
+        return None
+
+    print("\n***Port Deletion Test Passed***\n")
+
+    portJson = {
+        "PORT": {
+            "Ethernet0": {
```
Comment (praveen-li) [inline config/config_mgmt.py:561]: lanes, index and speed are mandatory. Change Yang model if needed.
admin_status not mandatory.

File: config/config_mgmt.py [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+    from sonic_yang import sonic_yang
+    import re
+
+except ImportError as e:
+    raise ImportError("%s - required module not found" % str(e))
+
+# Globals
+CONFIG_DB_JSON_FILE = '/etc/sonic/confib_db.json'
+# TODO: Find a place for it on sonic switch.
+DEFAULT_CONFIG_DB_JSON_FILE = 'default_config_db.json'
```
Comment (praveen-li) [inline config/config_mgmt.py]: default config will be in /etc/sonic. Default Config only contains ports.

File: config/config_mgmt.py [hunk 5]
```diff
@@ -0,0 +1,583 @@
...
+
+            self.DEBUG_FILE = None
+            if debug:
+                self.DEBUG_FILE = '_debug_config_mgmt'
+                with open(self.DEBUG_FILE, 'w') as df:
+                    df.write('--- Start config_mgmt logging ---\n\n')
+
+            # This class may not need to know about YANG_DIR ?, sonic_yang shd use
+            # default dir.
+            YANG_DIR = "/usr/local/yang-models"
```
Comment (praveen-li) [inline config/config_mgmt.py]: Move at the begining.

File: config/config_mgmt.py [hunk 6]
```diff
@@ -0,0 +1,583 @@
...
+            # this will crop config, xlate and load.
+            self.sy.load_data(self.configdbJsonIn)
+
+        except Exception as e:
+            print(e)
+            raise(Exception('configMgmt Class creation failed'))
+
+        return
+
+    def logInFile(self, header="", obj=None, json=False):
```
Comment (praveen-li) [inline config/config_mgmt.py:72]: Time stamp and size of file.

File: config/config_mgmt.py [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+CONFIG_DB_JSON_FILE = '/etc/sonic/confib_db.json'
+# TODO: Find a place for it on sonic switch.
+DEFAULT_CONFIG_DB_JSON_FILE = 'default_config_db.json'
+
+# Class to handle config managment for SONIC, this class will use PLY to verify
+# config for the commands which are capable of change in config DB.
+
+class configMgmt():
+
+    def __init__(self, source="configDBJson", debug=False):
```
Comment (praveen-li) [inline config/config_mgmt.py]: source = configDB

File: config/config_mgmt.py [hunk 8]
```diff
@@ -0,0 +1,583 @@
...
+    Add Ports and default config for ports to config DB, after validation of
+    data tree
+
+    PortJson: Config DB Json Part of all Ports same as PORT Table of Config DB.
+    ports = list of ports
+    force: If Force add default config as well.
+
+    return: Sucess: True or Failure: False
+    """
+    def addPorts(self, ports=list(), portJson=dict(), force=False):
```
Comment (praveen-li) [inline config/config_mgmt.py]: load default config. [Cli must have option for it.]

File: config/config_mgmt.py (hunk 3 above)
Comment (praveen-li) [reply config/config_mgmt.py:561]: Fixed by https://github.com/zhenggen-xu/sonic-buildimage/pull/19/commits

File: config/config_mgmt.py (hunk 1 above)
Comment (praveen-li) [reply config/config_mgmt.py:149]: Addressed in https://github.com/zhenggen-xu/sonic-buildimage/pull/19/files

**Review by zhenggen-xu — COMMENTED** [review]
In general, a few comments:
1,  We should have a function to make breakout atomic to configDB, I,E, we do all the checks etc and validation before we writing to configDB.

2, The default configuration check should also follow yang model.  We should probably have the dummy PORT section in default_configuration along with other configurations,  and we load the default configuration to internal yang tree, and then we merge the PORT trees that are relevant  to the working yang tree for the current configDB.

**Review by zhenggen-xu — APPROVED** [review]
Merge this for now to unblock other things.  Please address the comment in later commit.

**Pull request merged**